### PR TITLE
Fixing build on HIP Path

### PR DESCRIPTION
**Pull request by cjatin**
This Fixes build on HIP PATH.
Uses correct lib names and header files for newest HIP release.

### CODE REVIEW

**Comment by dev-zero** [comment]
ok to test

**Comment by alazzaro** [comment]
Thanks @cjatin for this PR!
We don't have a HIP test in place, could I ask @shoshijak to review this PR?

**Comment by dev-zero** [comment]
@cjatin if there are Docker images available with HIP libraries we could setup a compile-test as well.

**Comment by cjatin** [comment]
@dev-zero regarding docker, I believe you can use https://hub.docker.com/r/rocm/rocm-terminal/
PR is not final, more changes might come it. Currently trying to build unit tests.

**Comment by cjatin** [comment]
If you guys want a customized docker, the scripts can be found at https://github.com/RadeonOpenCompute/ROCm-docker/tree/master/dev
You'll also need hipblas, rocblas and rocsolver,  as well.

Need some help in getting the test up and running, Its failing while building tests:
```
[ 88%] Linking Fortran executable dbcsr_unittest1
cd /root/dbcsr/build/tests && /usr/bin/cmake -E cmake_link_script CMakeFiles/dbcsr_unittest1.dir/link.txt --verbose=1
/usr/bin/gfortran   -ffree-form -std=f2008ts -fimplicit-none -Werror=aliasing -Werror=ampersand -Werror=c-binding-type -Werror=intrinsic-shadow -Werror=intrinsics-std -Werror=line-truncation -Werror=tabs -Werror=target-lifetime -Werror=underflow -Werror=unused-but-set-parameter -Werror=unused-but-set-variable -Werror=unused-variable -Werror=unused-dummy-argument -Werror=conversion -Werror=zerotrip -Werror=uninitialized -Wno-maybe-uninitialized -Werror=argument-mismatch -O3 -g -funroll-loops CMakeFiles/dbcsr_unittest1.dir/dbcsr_unittest1.F.o  -o dbcsr_unittest1
-Wl,-rpath,/opt/rocm-3.7.0/hip/lib:/opt/rocm-3.7.0/lib libdbcsr_unittest_common.a ../src/libdbcsr.a /usr/lib/x86_64-linux-gnu/libblas.so /usr/lib/x86_64-linux-gnu/liblapack.so /usr/lib/x86_64-linux-gnu/libblas.so /usr/lib/x86_64-linux-gnu/liblapack.so ../src/acc/libsmm_acc/liblibsmm_acc.a /opt/rocm-3.7.0/hip/lib/libamdhip64.so ../src/libacc.a /opt/rocm-3.7.0/lib/libhipblas.so
/usr/bin/ld: ../src/acc/libsmm_acc/liblibsmm_acc.a(libsmm_acc_generated_libsmm_acc.cpp.o): undefined reference to symbol '_ZSt20__throw_length_errorPKc@@GLIBCXX_3.4'
/usr/bin/ld: /lib/x86_64-linux-gnu/libstdc++.so.6: error adding symbols: DSO missing from command line
collect2: error: ld returned 1 exit status
make[2]: *** [tests/CMakeFiles/dbcsr_unittest1.dir/build.make:94: tests/dbcsr_unittest1] Error 1
make[2]: Leaving directory '/root/dbcsr/build'
make[1]: *** [CMakeFiles/Makefile2:1253: tests/CMakeFiles/dbcsr_unittest1.dir/all] Error 2
make[1]: Leaving directory '/root/dbcsr/build'
make: *** [Makefile:141: all] Error 2
```


CMake Config: 
`cmake .. -DCMAKE_INSTALL_PREFIX=$PWD/install -DUSE_HIP=ON -DWITH_EXAMPLES=OFF -DWITH_GPU=Mi50 -DUSE_SMM=blas -DBUILD_TESTING=ON -DUSE_OPENMP=OFF -DUSE_MPI=OFF`

**Comment by alazzaro** [comment]
Well, this is
```
 > c++filt _ZSt20__throw_length_errorPKc
std::__throw_length_error(char const*)
```
It seems there is an order problem of the library or something broken with the C++ compilation part.
Could you check that your C++ compiler is GCC? I wonder if to have clang installed... (just guessing, no real clue)

**Comment by dev-zero** [comment]
we probably have to explicitly add a `-lstdc++` when linking with the Fortran compiler

**Review by shoshijak — CHANGES_REQUESTED** [review]
Thanks a lot for the fixes to the hipblas interface!

**Comment by shoshijak** [comment]
> @cjatin if there are Docker images available with HIP libraries we could setup a compile-test as well.

Yes, we should be able to do that and it would be a good thing for the HIP backend. I see [SIRIUS](https://github.com/electronic-structure/SIRIUS/blob/master/.github/workflows/build.yml), for example, have a compile-test running with github actions

File: CMakeLists.txt [hunk 1]
```diff
@@ -222,7 +222,7 @@ if (USE_HIP)
   endif()
 
   # Find hiprtc library (adds support for JIT-ing in HIP)
-  find_library(ROCM_HIPRTC_LIB hiprtc HINTS ${HIP_PATH}/lib)
+  find_library(ROCM_HIPRTC_LIB amdhip64 HINTS ${HIP_PATH}/lib)
```
[the remark this answers is not in the document]
Comment (shoshijak) [reply CMakeLists.txt:225]: I now see that `libhiprtc.so` symlinks to `libamdhip64.so` on my HIP installation, but this is a recent change in HIP if I'm not mistaken ... @cjatin , do you know what is best practice currently?

[the remark this answers is not in the document]
Comment (cjatin) [reply CMakeLists.txt:225]: Yes this change was recent when hcc was deprecated, `libamdhip64.so` should be used going ahead.

**Comment by cjatin** [comment]
> > @cjatin if there are Docker images available with HIP libraries we could setup a compile-test as well.
> 
> Yes, we should be able to do that and it would be a good thing for the HIP backend. I see [SIRIUS](https://github.com/electronic-structure/SIRIUS/blob/master/.github/workflows/build.yml), for example, have a compile-test running with github actions

I have provided the docker script which can be used to create a docker image.

> If you guys want a customized docker, the scripts can be found at https://github.com/RadeonOpenCompute/ROCm-docker/tree/master/dev
> You'll also need hipblas, rocblas and rocsolver, as well.

**Comment by shoshijak** [comment]
Hi @cjatin , I fixed the build on the HIP path and took the liberty of pushing directly to this PR, hope you don't mind. 
I'll look into a a CI compile-test now so we can prevet such breaks in the future.

**Comment by cjatin** [comment]
> Hi @cjatin , I fixed the build on the HIP path and took the liberty of pushing directly to this PR, hope you don't mind.
> I'll look into a a CI compile-test now so we can prevent such breaks in the future.

Thanks.
Let me know if you need any information.

**Comment by cjatin** [comment]
I tried checking out the PR with new changes. I still get the following error, there is '-lstdc++' now.


```
[ 87%] Linking Fortran executable dbcsr_unittest2
cd /root/driver/dbcsr/build/tests && /usr/bin/cmake -E cmake_link_script CMakeFiles/dbcsr_unittest2.dir/link.txt --verbose=1
/usr/bin/gfortran  -lstdc++ -ffree-form -std=f2008ts -fimplicit-none -Werror=aliasing -Werror=ampersand -Werror=c-binding-type -Werror=intrinsic-shadow -Werror=intrinsics-std -Werror=line-truncation -Werror=tabs -Werror=target-lifetime -Werror=underflow -Werror=unused-but-set-parameter -Werror=unused-but-set-variable -Werror=unused-variable -Werror=unused-dummy-argument -Werror=conversion -Werror=zerotrip -Werror=uninitialized -Wno-maybe-uninitialized -Werror=argument-mismatch -O2 -ggdb -fno-omit-frame-pointer -fcheck=all -ffpe-trap=invalid,zero,overflow -fbacktrace -finit-real=snan -finit-integer=-42 -finit-derived -Werror=realloc-lhs -finline-matmul-limit=0 -fsanitize=undefined -fsanitize=address -fsanitize-recover=all -Warray-temporaries -fsanitize=leak CMakeFiles/dbcsr_unittest2.dir/dbcsr_unittest2.F.o  -o dbcsr_unittest2  -Wl,-rpath,/opt/rocm-3.9.0-3629/hip/lib:/opt/rocm-3.9.0-3629/lib libdbcsr_unittest_common.a ../src/libdbcsr.a /usr/lib/x86_64-linux-gnu/libopenblas.so -lpthread -lm -ldl ../src/acc/libsmm_acc/liblibsmm_acc.a /opt/rocm-3.9.0-3629/hip/lib/libamdhip64.so ../src/libacc.a /opt/rocm-3.9.0-3629/lib/libhipblas.so
/usr/bin/ld: ../src/acc/libsmm_acc/liblibsmm_acc.a(libsmm_acc_generated_libsmm_acc.cpp.o): undefined reference to symbol '_ZSt9terminatev@@GLIBCXX_3.4'
/usr/lib/gcc/x86_64-linux-gnu/7/libstdc++.so: error adding symbols: DSO missing from command line

```

**Comment by dev-zero** [comment]
> I tried checking out the PR with new changes. I still get the following error, there is '-lstdc++' now. The error signature seems to be the same. Am i missing a dependency?
> 
> ```
> [ 87%] Linking Fortran executable dbcsr_unittest2
> cd /root/driver/dbcsr/build/tests && /usr/bin/cmake -E cmake_link_script CMakeFiles/dbcsr_unittest2.dir/link.txt --verbose=1
> /usr/bin/gfortran  -lstdc++ -ffree-form -std=f2008ts -fimplicit-none -Werror=aliasing -Werror=ampersand -Werror=c-binding-type -Werror=intrinsic-shadow -Werror=intrinsics-std -Werror=line-truncation -Werror=tabs -Werror=target-lifetime -Werror=underflow -Werror=unused-but-set-parameter -Werror=unused-but-set-variable -Werror=unused-variable -Werror=unused-dummy-argument -Werror=conversion -Werror=zerotrip -Werror=uninitialized -Wno-maybe-uninitialized -Werror=argument-mismatch -O2 -ggdb -fno-omit-frame-pointer -fcheck=all -ffpe-trap=invalid,zero,overflow -fbacktrace -finit-real=snan -finit-integer=-42 -finit-derived -Werror=realloc-lhs -finline-matmul-limit=0 -fsanitize=undefined -fsanitize=address -fsanitize-recover=all -Warray-temporaries -fsanitize=leak CMakeFiles/dbcsr_unittest2.dir/dbcsr_unittest2.F.o  -o dbcsr_unittest2  -Wl,-rpath,/opt/rocm-3.9.0-3629/hip/lib:/opt/rocm-3.9.0-3629/lib libdbcsr_unittest_common.a ../src/libdbcsr.a /usr/lib/x86_64-linux-gnu/libopenblas.so -lpthread -lm -ldl ../src/acc/libsmm_acc/liblibsmm_acc.a /opt/rocm-3.9.0-3629/hip/lib/libamdhip64.so ../src/libacc.a /opt/rocm-3.9.0-3629/lib/libhipblas.so
> /usr/bin/ld: ../src/acc/libsmm_acc/liblibsmm_acc.a(libsmm_acc_generated_libsmm_acc.cpp.o): undefined reference to symbol '_ZSt9terminatev@@GLIBCXX_3.4'
> /usr/lib/gcc/x86_64-linux-gnu/7/libstdc++.so: error adding symbols: DSO missing from command line
> ```

AFAIK `-lstdc++` is in the wrong order (IF the missing `-lstdc++` is indeed the problem): most toolchains nowadays have an implicit `-Wl,--as-needed`, making the linker strip unused symbols incrementally in order of their appearance on the command line, meaning that when it starts linking `src/acc/libsmm_acc/liblibsmm_acc.a(libsmm_acc_generated_libsmm_acc.cpp.o` the unused symbols from `libstdc++` have already been stripped. So, `-lstdc++` should go after `../src/acc/libsmm_acc/liblibsmm_acc.a` to have any effect.

**Comment by dev-zero** [comment]
... so, I would say the solution is to use `target_link_libraries(libsmm_acc INTERFACE "stdc++")` in `src/acc/libsmm_acc/CMakeLists.txt ` rather than changing the linker options just for the tests. Since it is a static library, using `INTERFACE` should be the right thing, but if it doesn't work, leave the `INTERFACE` keyword away completely or use `PUBLIC` instead.

@cjatin can you please update the HIP build system to support `OBJECT` libraries in `hip_add_library`?

**Comment by cjatin** [comment]
> ... so, I would say the solution is to use `target_link_libraries(libsmm_acc INTERFACE "stdc++")` in `src/acc/libsmm_acc/CMakeLists.txt ` rather than changing the linker options just for the tests. Since it is a static library, using `INTERFACE` should be the right thing, but if it doesn't work, leave the `INTERFACE` keyword away completely or use `PUBLIC` instead.
> 
> @cjatin can you please update the HIP build system to support `OBJECT` libraries in `hip_add_library`?

are you looking for something like this?
https://github.com/ROCm-Developer-Tools/HIP/blob/cb1a3bb60f1c89c6967eefb432f4762182a113f2/cmake/FindHIP.cmake#L562

**Comment by shoshijak** [comment]
Okay, we have a compile test for the HIP backend 🎉
@dev-zero , would you mind taking a look whether this seems sounds to you?

File: .github/workflows/testing-linux.yml [hunk 2]
```diff
@@ -75,4 +76,34 @@ jobs:
         yml: .codecov.yml
         fail_ci_if_error: true
 
+  ##################################################################################
+  # Build on ROCm
+  ##################################################################################
+  build-on-rocm:
+    runs-on: ubuntu-latest
+    container:
+      image: shoshijak/build-env-rocm:1.0
```
Comment (shoshijak) [inline .github/workflows/testing-linux.yml:85]: The Dockerfile I wrote should be pushed to dbcsr's dockerhu, and not stay on mine

**Review by shoshijak — APPROVED** [review]

**Pull request merged**